### PR TITLE
D3 alignment tests accommodating heat map

### LINKED ISSUE #71: D3 Heat Map - Test #9 Edge Case
#### Issue Description
I raised a similar issue yesterday with the scatter plot project, but the same problem appears to be happening with the heat map project I built today, too. It's failing a test because apparently the cells don't line up with the appropriate y-axis values, when they clearly do just by looking at the graph, or by checking out the attributes of the relevant svg elements using Chrome's dev tools. Might be worth having a look at that test and maybe seeing if there is a more flexible way of checking if a project has satisfied the particular user story. The current test logic is a bit too restrictive IMO!

#### Your Code / Link to Your Pen

Codepen: http://codepen.io/tom-p-uk/pen/ZeQEaE?editors=1010

Failing test code:

```
 it('9. My heat map should have cells that align with the corresponding month on the y-axis.', function () {
	                var cellsCollection = document.querySelectorAll('.cell');
	                FCC_Global.assert.isAbove(cellsCollection.length, 0, "Could not find any elements with a class=\"cell\" ");

	                //convert to array
	                var cells = [].slice.call(cellsCollection);
	                var sortedCells = cells.sort(function (a, b) {
	                    return a.getAttribute("data-month") - b.getAttribute("data-month");
	                });

	                //check to see if the y locations of the new sorted array are in ascending order
	                for (var i = 0; i < sortedCells.length - 1; ++i) {
	                    FCC_Global.assert.isAtMost(+sortedCells[i].getAttribute("y"), +sortedCells[i + 1].getAttribute("y"), "month values don't line up with y locations ");
	                }
	            });
```

### PR DESCRIPTION
**Pull request by tbushman**
#### Pre-Submission Checklist

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file
- [x] Working changes have been added to the build by running `npm run build`

#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.

- [x] Closes currently open issue: Closes #71 

#### Description

Changed alignmentD3Tests.js to accommodate rect and months. Changes necessitated changing scatter-plot-tests.js as well.

### CODE REVIEW

File: src/assets/alignmentD3Tests.js [hunk 1]
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
...
+        count++;
       }
+    }
   }
   return count;
 };
 
 /**
- * Get x coordinate of given feature
+ * Get x or y coordinate of given feature
```
Comment (tbushman) [inline src/assets/alignmentD3Tests.js]: revert comment

File: src/project-tests/heat-map-tests.js [hunk 2]
```diff
@@ -168,21 +179,25 @@ export default function createHeatMapTests() {
...
-        // check to see if the y locations of the new sorted array are in
-        // ascending order
-        for (var i = 0; i < sortedCells.length - 1; ++i) {
-          FCC_Global.assert.isAtMost(
-            +sortedCells[i].getAttribute('y'),
-            +sortedCells[i + 1].getAttribute('y'),
-            'month values don\'t line up with y locations '
-          );
-        }
+        FCC_Global.assert.isTrue(
```
Comment (tbushman) [inline src/project-tests/heat-map-tests.js]: needs descriptive comment

File: src/assets/alignmentD3Tests.js [hunk 3]
```diff
@@ -124,37 +137,77 @@ var getXTickLocation = function(tick) {
...
+export function getFeatureValueInteger(feature, dataAttr) {
+  var value = feature.getAttribute(dataAttr);
 
   return parseInt(value, 10);
 };
 
 /**
+ * Get feature data-* attribute
+ * @function
+ * @param {String} feature - HTMLElement rect (.cell)
```
Comment (tbushman) [inline src/assets/alignmentD3Tests.js]: `@param {HTMLElement} feature`

File: src/assets/alignmentD3Tests.js [hunk 4]
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
...
   }
   return count;
 };
 
 /**
- * Get x coordinate of given feature
+ * Get x or y coordinate of given feature
  * @function
- * @param {String} feature - HTMLElement circle (.dot)
+ * @param {String} feature - HTMLElement (circle or rect)
```
Comment (tbushman) [inline src/assets/alignmentD3Tests.js]: `@param {HTMLElement} feature`

File: src/assets/alignmentD3Tests.js [hunk 5]
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
...
-      if (featureVal >= tickValCur && featureVal <= tickValNext) {
-          if (!isFeatureAligned(featureCoord, axis, tickPxCur, tickPxNext)) {
-            count++;
-          }
+    // get values for given feature (j)
+    var featureVal = getFeatureValueFunc(collection.item(j), axis.dataAttr),
+      featureCoord = getFeatureCoord(collection.item(j), axis.coordAttr);
+    if (featureVal >= tickValCur && featureVal < tickValNext) {
+      if (!isFeatureAligned(featureCoord, axis, tickPxCur, tickPxNext)) {
+        console.log(collection.item(j), axis.coordAttr, featureCoord, tickPxCur, tickPxNext)
```
Comment (tchaffee) [inline src/assets/alignmentD3Tests.js]: `console.log` should be removed, unless it serves a production purpose. This looks like debugging?

File: src/assets/alignmentD3Tests.js [hunk 6]
```diff
@@ -124,37 +137,77 @@ var getXTickLocation = function(tick) {
...
+ * @returns {Number} integer
+ */
+export function getTickValueInteger(axisTick) {
+  return parseInt(axisTick, 10);
+};
+
+/**
+ * Convert one axis tick's label innerHTML from month string to index
+ * @function
+ * @param {String} axisTick - innerHTML from text of a single tick. Ex: "March"
```
Comment (tchaffee) [inline src/assets/alignmentD3Tests.js]: This function assumes the time being formatted as %B (see https://github.com/d3/d3-3.x-api-reference/blob/master/Time-Formatting.md) which is "full month name". This is included in the heat map tests (see test #<span>11 on line 234 in file heat-map-tests.js). I think this would be more  reliable if you declare the `months` const only in heat-map-tests.js and pass it as a parameter to `isAxisAlignedWithDataPoints`. This is a case of DRY since the `months` const is declared in two places and should be declared in only one place.

File: src/assets/alignmentD3Tests.js [hunk 7]
```diff
@@ -314,17 +353,29 @@ var getMisalignmentCountCaller = function(
  * @param {Array}  collection - HTMLCollection of features
```
Comment (tchaffee) [inline src/assets/alignmentD3Tests.js]: What does an HTMLCollection of features look like? Since this is the main function called from the tests, it should be well documented what the parameters look like. A code example of how you would get such a collection would be helpful. For example, in the heat map tests I think the code is `const cellsCollection = document.querySelectorAll('.cell');`? That code sample, along with an explanation that it's getting all the D3 data points, would be helpful to understand what the param is supposed to be.

File: src/assets/alignmentD3Tests.js [hunk 8]
```diff
@@ -314,17 +353,29 @@ var getMisalignmentCountCaller = function(
...
  * @param {Function} getMisalignmentCountFunc -
  * passes either getXMisalignmentCount or getYMisalignmentCount
+ * @param {Function} getFeatureValueFunc - getFeatureValueInteger or
+ * getFeatureValueMinutes
+ * @param {Function} getTickValueFunc - getTickValueInteger,
+ * getTickValueMinutes, or getTickValueMonths
  * @returns {Boolean} true if no misalignments are counted
```
Comment (tchaffee) [inline src/assets/alignmentD3Tests.js]: It is a minor detail, but I always put a space before the `@returns` so it stands out from the parameters.

**Comment by tbushman** [comment]
@tchaffee OK, I added an optional parameter for hard-coded axis values like months in alignmentD3Tests.getAxisInfo and addressed your other feedback, fixed comments, etc., and tested locally and in codepen with rawgit bundle.js. I've removed the blocked tag and this is ready to merge.

**Comment by tchaffee** [comment]
LGTM
